### DISCUSSION TITLE
Prisma 2.5 middleware/subscription examples

### DISCUSSION THREAD

**Discussion by TheoMer** [prisma/prisma#3367] [Q&A] [answered]
Hey guys, really enjoyed your live YouTube presentation on 2.5 earlier today.

In regards to the issue of subscriptions that was spoken on, do you have any middleware/subscription examples you could point me to?

**Comment by ryands17** [comment, marked as answer]
Hey @TheoMer 👋 
I don't think there's any combination of middleware and subscriptions, but we have a basic [subscription example](https://github.com/prisma/prisma-examples/tree/master/typescript/subscriptions-pubsub) and also a community member has created [this](https://www.npmjs.com/package/prisma-lrucache-middleware) with middlewares.

**Reply by TheoMer** [reply]
Hey @ryands17 Many thanks.

**Comment by TheoMer** [comment]
Hey @ryands17 so I followed the subscription example and have got subscriptions working. What I am deliberating is how to replicate the mutation check parameter that was part of Prisma 1:

```ts
const ALL_ITEMS_SUBSCRIPTION = gql`
  subscription {
    item {
      mutation
      node {
        id        
      }
    }
  }
`;
```

In the Prisma 1 version of my primary subscription I do a mutationType check:

```ts
        subscribeToNewItems={() =>
          subscribeToMore({
            document: ALL_ITEMS_SUBSCRIPTION,
            variables: {},
            updateQuery: (prev, { subscriptionData }) => {
              console.log("SubscriptionWatch in Items.js = ", subscriptionData.data);
              if (!subscriptionData.data) return prev;

              const newItem = subscriptionData.data.item.node;
              const mutationType = subscriptionData.data.item.mutation;

              // Check that the item doesn't already exist in the store
              if (prev.items.find((item) => item.id === newItem.id)) {
                return prev;                    
              }

              if (mutationType === 'CREATED') {
                return Object.assign({}, prev, {
                  items: [newItem, ...prev.items]
                });
              }
            }
          })
        }
```

But I'm not sure, short of adding a mutation column to the item table, how to get pubsub to replicate this.

So in 3 mutation resolvers I do:

deleteItem
```ts
ctx.pubsub.publish('itemWatch', deletedItem);
```

createOneItem
```ts
ctx.pubsub.publish('itemWatch', newItem);
```

updateOneItem
```ts
ctx.pubsub.publish('itemWatch', updateItem);
```

which I handle:

```ts
export const ItemWatch = subscriptionField('itemWatch', {
    type: 'Item',
    subscribe(_root, _args, ctx) {
      const { pubsub } = ctx;
      return pubsub.asyncIterator('itemWatch');
    },
    resolve(payload) {
      return payload
    },
});
```

So how with my current subscription modification can I check what type of mutation has taken place?:

```ts
const ALL_ITEMS_SUBSCRIPTION = gql`
  subscription {
    itemWatch {
      id        
    }
  }
`;
```

You can find my current repo here: https://github.com/TheoMer/Prisma2-Miration

**Reply by ryands17** [reply]
@TheoMer You could do it like this. I am using `deleteItem` as an example.


deleteItem

```ts
ctx.pubsub.publish('itemWatch', { type: 'delete', item: deletedItem });
```

And in your subscription:

```ts
export const ItemWatch = subscriptionField('itemWatch', {
    type: 'Item',
    subscribe(_root, _args, ctx) {
      const { pubsub } = ctx;
      return pubsub.asyncIterator('itemWatch');
    },
    resolve(payload) {
      console.log(payload.type) // here you will get the operation
      return payload.item
    },
});
```

**Reply by TheoMer** [reply]
Hey @ryands17 Many thanks for the heads up. Can you clarify how I make payload.type visible to the client? I tried:

```
      const newItem = [...payload.item];
      newItem.push({
        mutation: payload.type,
      });
     return newItem

```

without success.

**Reply by ryands17** [reply]
You would need to create a new object:

```ts
export const Post = objectType({
  name: 'ItemModifier',
  definition(t) {
    t.field({ type: 'Item' })
    t.string('type')
  },
})
```

And in your subscription:

```ts
export const ItemWatch = subscriptionField('itemWatch', {
    type: 'ItemModifier',
    subscribe(_root, _args, ctx) {
      const { pubsub } = ctx;
      return pubsub.asyncIterator('itemWatch');
    },
    resolve(payload) {
      return payload
    },
});
```

**Reply by TheoMer** [reply]
@ryands17 You are an absolute diamond. Many thanks.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
